pkg/scheduler/core/generic_scheduler.go

```diff
@@ -479,20 +478,17 @@ func (g *genericScheduler) findNodesThatFit(ctx context.Context, state *framewor
 		state.Write(migration.PredicatesStateKey, &migration.PredicatesStateData{Reference: meta})
 
 		checkNode := func(i int) {
-			nodeName := g.cache.NodeTree().Next()
-
+			nodeInfo := g.nodeInfoSnapshot.NodeInfoList[i]
```

Comment:
This broke scalability tests: https://github.com/kubernetes/kubernetes/issues/84151

TL;DR; this breaks spreading of pods in large clusters.

What exactly has happened:
1, in large enough clusters, we are using the features of finding only N feasible nodes and scoring only those:
https://github.com/kubernetes/kubernetes/blob/9d173852c14f0e8efb0d67db2c38b8dcfa45b31b/pkg/scheduler/core/generic_scheduler.go#L464

2. Before this change, we were looking for nodes starting at the point where we previously stopped (because Next() was done at the level of the original tree).
3. With this change, we're always starting from 0.

So assume, you have 5k nodes, all are feasible, and numFeasible is choosing 250.
- previously, you first chose nodes [1..250], then [251..500], ... [4751..5000], [1..250], ...
- with this change, we will chose [1..250], [1..250], [1..250], ... until one of those nodes become unfeasible and we will choose something different.

With this PR, next() is called only in UpdateNodeSnapshotInfo:
https://github.com/kubernetes/kubernetes/pull/84014/files#diff-f4a894ca5e905aa5f613269fc967fe2cR206
and if the set of nodes doesn't change, we will pretty much always be generating the same set of nodes.

This kind of breaks the fact that scheduler is scheduling in the whole cluster. While it's not documented feature per-se, I think this isn't the right thing to do.

I'm going to open a revert of this PR to to fix scalability tests (or the half of them, because we seem to have two different regressions), but will wait for your explicit approval.
We can discuss how to fix that later.

Reply:
heh... it's no longer possible to autorevert it..

Reply:
Fortunately the conflicts were trivial - opened https://github.com/kubernetes/kubernetes/pull/84222